allensdk/internal/brain_observatory/time_sync.py

```diff
@@ -18,37 +18,49 @@
 LONG_STIM_THRESHOLD = 0.2     # seconds
 MAX_MONITOR_DELAY = 0.07     # seconds
 
-VERSION_1_KEYS = {
-    "photodiode": "stim_photodiode",
-    "2p": "2p_vsync",
-    "stimulus": "stim_vsync",
-    "eye_camera": "cam2_exposure",
-    "behavior_camera": "cam1_exposure",
-    "acquiring": "2p_acquiring",
-    "lick_sensor": "lick_1"
-    }
-
-# MPE is changing keys. This isn't versioned in the file.
-VERSION_2_KEYS = {
-    "photodiode": "photodiode",
-    "2p": "2p_vsync",
-    "stimulus": "stim_vsync",
-    "eye_camera": "eye_tracking",
-    "behavior_camera": "behavior_monitoring",
-    "acquiring": "2p_acquiring",
-    "lick_sensor": "lick_sensor"
-    }
-
-
-def get_keys(sync_dset):
-    """Get the correct lookup for line labels.
-
-    This method is fragile, but not all old data contains the full list
-    of keys.
+
+def get_keys(sync_dset: Dataset) -> dict:
+    """
+    Gets the correct keys for the sync file by searching the sync file
+    line labels. Assumes to use old keys if none of the correct ones are
+    found in line labels
+    Args:
+        sync_dset: The sync dataset to search for keys within
+
+    Returns:
+        key_dict: dictionary of key value pairs for finding data in the
+                  sync file
     """
-    if "cam2_exposure" in sync_dset.line_labels:
-        return VERSION_1_KEYS
-    return VERSION_2_KEYS
+    key_dict = {
```

Comment:
I think the `VERSION_1_KEYS`, `VERSION_2_KEYS` should be kept, also for the new failures a `VERSION_3_KEYS` should be created. I think there is value in explicitly documenting specific combinations of keys rather than trying a key and seeing what works. If sync files someday in the future, accidentally have multiple valid keys, or the schema suddenly changes, I don't think it's a good idea to just silently figure out what keys it has and keep going as this may allow more insidious sync file errors (e. g. maybe we end up reading data from the wrong key) to creep by without any notice.

So I think the `get_keys` function should probably work more like:

1. Determine which 'version' a sync file's keys are
2. Return the keys for that version
3. Throw an error if a sync file has a combination of keys that don't match any known versions

I'm curious as to what others think?

Reply:
That was my thought originally as well. I moved away from this because I didn't like inferring the keys mapped perfectly to one version. The error that caused this happened because the 'photodiode' key was different in actuality than what was reported in the version it mapped to. This to me has a higher danger moving forward. The scientists are free to essentially change keys file to file and very easily break these versions without us knowing at all, that's what broke the original test case. I think it should be based off the keys present in the line labels without inference. This preserves usability of the files if they don't have perfect key mappings whereas the subsequent code would break otherwise.